streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/ApplicationStateImpl.java

```diff
@@ -0,0 +1,76 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.streams.processor.internals.assignment;
+
+import static java.util.Collections.unmodifiableMap;
+import static java.util.Collections.unmodifiableSet;
+
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+import org.apache.kafka.streams.processor.TaskId;
+import org.apache.kafka.streams.processor.assignment.ApplicationState;
+import org.apache.kafka.streams.processor.assignment.AssignmentConfigs;
+import org.apache.kafka.streams.processor.assignment.KafkaStreamsState;
+import org.apache.kafka.streams.processor.assignment.ProcessId;
+
+public class ApplicationStateImpl implements ApplicationState {
+
+    private final AssignmentConfigs assignmentConfigs;
+    private final Set<TaskId> statelessTasks;
+    private final Set<TaskId> statefulTasks;
+    private final Map<ProcessId, KafkaStreamsState> kafkaStreamsStates;
+
+    public ApplicationStateImpl(
+        final AssignmentConfigs assignmentConfigs,
+        final Map<ProcessId, KafkaStreamsState> kafkaStreamsStates,
+        final Set<TaskId> statefulTasks,
+        final Set<TaskId> statelessTasks
+    ) {
+        this.assignmentConfigs = assignmentConfigs;
+        this.kafkaStreamsStates = unmodifiableMap(kafkaStreamsStates);
+        this.statefulTasks = unmodifiableSet(statefulTasks);
+        this.statelessTasks = unmodifiableSet(statelessTasks);
+    }
+
+    @Override
+    public Map<ProcessId, KafkaStreamsState> kafkaStreamsStates(final boolean computeTaskLags) {
+        return kafkaStreamsStates;
```

Comment:
We should make sure to respect the `computeTaskLags` param here. There's two things we can do: wait until this method and then call some `KafkaStreamsStateImpl#computeTaskLags` method, or just wait to construct the KafkaStreamsImpls at all until we know whether or not we should compute the task lags, and then pass the `computeTaskLags` flag into the KafkaStreamsImpl constructor.

I personally prefer the latter since that way there's no partially-initialized classes floating around and we don't have to keep track of when the task lags are computed/initialized.